openwisp_controller/config/api/views.py

```diff
@@ -125,6 +132,114 @@ class DeviceGroupDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+    return url
+
+
+class DeviceGroupCommonName(ProtectedAPIMixin, RetrieveAPIView):
+    serializer_class = DeviceGroupSerializer
+    queryset = DeviceGroup.objects.select_related('organization').order_by('-created')
+    # Not setting lookup_field makes DRF raise error. but it is not used
+    lookup_field = 'pk'
+
+    @classmethod
+    @cache_memoize(
+        timeout=24 * 60 * 60, args_rewrite=get_cached_devicegroup_args_rewrite
+    )
+    def get_device_group(cls, org_slugs, common_name):
+        query = Q(common_name=common_name)
+        if org_slugs:
+            org_slugs = org_slugs.split(',')
+            query = query & Q(organization__slug__in=org_slugs)
+        try:
+            cert = (
+                Cert.objects.select_related('organization')
+                .only('id', 'organization')
+                .filter(query)
+                .first()
+            )
+            vpnclient = VpnClient.objects.only('config_id').get(cert_id=cert.id)
+            group = (
+                Device.objects.select_related('group')
+                .only('group')
+                .get(config=vpnclient.config_id)
+                .group
+            )
+            assert group is not None
+        except (ObjectDoesNotExist, AssertionError, AttributeError):
+            raise Http404
+        return group
+
+    def get_object(self):
+        org_slugs = self.request.query_params.get('org', '')
+        common_name = self.kwargs['common_name']
+        group = self.get_device_group(org_slugs, common_name)
+        # May raise a permission denied
+        self.check_object_permissions(self.request, group)
+        return group
+
+    @classmethod
+    def _invalidate_from_queryset(cls, queryset):
+        for obj in queryset.iterator():
+            cls.get_device_group.invalidate(None, '', obj['common_name'])
+            cls.get_device_group.invalidate(
+                None, obj['organization__slug'], obj['common_name']
+            )
+
+    @classmethod
+    def devicegroup_change_invalidates_cache(cls, device_group_id):
+        qs = (
+            VpnClient.objects.select_related(
+                'config',
+                'config__device',
+                'config__device__group',
+                'organization',
+                'cert',
+            )
+            .filter(config__device__group_id=device_group_id)
+            .annotate(
+                organization__slug=F('cert__organization__slug'),
+                common_name=F('cert__common_name'),
+            )
+            .values('common_name', 'organization__slug')
+        )
+        cls._invalidate_from_queryset(qs)
+
+    @classmethod
+    def organization_change_invalidates_cache(cls, organization_id):
+        qs = (
+            Cert.objects.select_related('organization')
+            .filter(organization_id=organization_id)
+            .values('organization__slug', 'common_name')
+        )
+        cls._invalidate_from_queryset(qs)
```

Comment:
so as discussed we can probably remove this logic because it's not super common and not having it will not be a big deal (new slugs get cached, old cached content gets expired).